[BUG] panic in qson package

The bug is found using the qson-fuzzer from [this PR](https://github.com/micro/go-micro/pull/1873).

The stack trace:
```
panic: runtime error: slice bounds out of range [6:5]

goroutine 17 [running, locked to thread]:
github.com/micro/go-micro/util/qson.buildNewKey(0x10c0000c58f0, 0x6, 0x6, 0xffffffffffffffff)
        github.com/micro/go-micro/util/qson/qson.go:141 +0x111
github.com/micro/go-micro/util/qson.queryToMap(0x10c0000c58f0, 0x7, 0x695641, 0x1, 0x0)
        github.com/micro/go-micro/util/qson/qson.go:118 +0x45a
github.com/micro/go-micro/util/qson.ToJSON(0x10c0000c58f0, 0x7, 0x7, 0x7, 0x10c0000c58f0, 0x7, 0x10c000044e80)
        github.com/micro/go-micro/util/qson/qson.go:58 +0x12e
github.com/micro/go-micro/util/qson.Fuzz(...)
        github.com/micro/go-micro/util/qson/qson-fuzz.go:4
main.LLVMFuzzerTestOneInput(0x60200000ebb0, 0x7, 0x18)
        command-line-arguments/main.650121307.go:21 +0x7b
main._cgoexpwrap_5519e206b9ee_LLVMFuzzerTestOneInput(0x60200000ebb0, 0x7, 0x9f5200)
        _cgo_gotypes.go:53 +0x46
```

Input:
`[^@hGl5=`

The environment was setup by cloning go-micro's repository, placing the fuzzer in the relevant directory and running the fuzzer.

--- a/util/qson/qson.go
+++ b/util/qson/qson.go
@@ -25,6 +25,19 @@ func init() {
 	bracketSplitter = regexp.MustCompile("\\[|\\]")
 }
 
+func btSplitter(str string) []string {
+	r := bracketSplitter.Split(str, -1)
+	for idx, s := range r {
+		if len(s) == 0 {
+			if len(r) > idx+1 {
+				copy(r[idx:], r[idx+1:])
+				r = r[:len(r)-1]
+			}
+		}
+	}
+	return r
+}
+
 // Unmarshal will take a dest along with URL
 // query params and attempt to first turn the query params
 // into JSON and then unmarshal those into the dest variable
@@ -86,7 +99,7 @@ func queryToMap(param string) (map[string]interface{}, error) {
 		return nil, err
 	}
 
-	pieces := bracketSplitter.Split(rawKey, -1)
+	pieces := btSplitter(rawKey)
 	key := pieces[0]
 
 	// If len==1 then rawKey has no [] chars and we can just
@@ -136,7 +149,8 @@ func queryToMap(param string) (map[string]interface{}, error) {
 // pieces = [bar one two ]
 // and return "one[two]"
 func buildNewKey(origKey string) string {
-	pieces := bracketSplitter.Split(origKey, -1)
+	pieces := btSplitter(origKey)
+
 	ret := origKey[len(pieces[0])+1:]
 	ret = ret[:len(pieces[1])] + ret[len(pieces[1])+1:]
 	return ret

--- a/util/qson/qson_test.go
+++ b/util/qson/qson_test.go
@@ -5,6 +5,14 @@ import (
 	"testing"
 )
 
+func TestFuzz1(t *testing.T) {
+	b, err := ToJSON(`[^@hGl5=`)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_ = b
+}
+
 func ExampleUnmarshal() {
 	type Ex struct {
 		A string `json:"a"`

